# Puzzle 2

--- Day 2: Dive! ---

Ahora, debes descubrir cómo pilotar esta cosa.

Parece que el submarino puede tomar una serie de comandos como avanzar 1, bajar 2 o subir 3:

forward X aumenta la posición horizontal en X unidades.
down X aumenta la profundidad en X unidades.
up X disminuye la profundidad en X unidades.
Tenga en cuenta que, dado que está en un submarino, subir y bajar afectan su profundidad, por lo que tienen el resultado opuesto de lo que podría esperar.

Parece que el submarino ya tiene un rumbo planificado (tu entrada del rompecabezas). Probablemente debería averiguar hacia dónde se dirige. Por ejemplo:

forward 5

down 5

forward 8

up 3

down 8

forward 2

Su posición horizontal y profundidad comienzan en 0. Los pasos anteriores los modificarían de la siguiente manera:

forward 5 suma 5 a su posición horizontal, un total de 5.
down 5 agrega 5 a su profundidad, lo que resulta en un valor de 5.
forward 8 suma 8 a su posición horizontal, un total de 13.
up 3 disminuye la profundidad en 3, lo que da como resultado un valor de 2.
down 8 agrega 8 a su profundidad, lo que resulta en un valor de 10.
forward 2 suma 2 a su posición horizontal, un total de 15.
Después de seguir estas instrucciones, tendrías una posición horizontal de 15 y una profundidad de 10. (Multiplicarlos juntos produce 150).

Calcula la posición horizontal y la profundidad que tendrías después de seguir el rumbo planificado. ¿Qué obtienes si multiplicas tu posición horizontal final por tu profundidad final?

## Logica de Resolucion

En este caso vemos que lo que quereos es obtener la posicion de como se mueve nuestro submarino, y la entradas de dat trae tanto una instruccion sobre como movernos, como tambien una cantidad de movimiento, para eso:

* Recorremos lo valores de nuestra entrada de datos, con un ciclo for

* el mensaje que lo recibimos lo vamos a separar en `comando` que sera la instruccion de como moverse y tambien la `cantidad` que guardara el valor de la cantidad de movimiento que tiene que hacer.

* luego veremos el valor que tiene el comando y de acuerdo a eso vamos a cambiar los valores de `posicion` y de `profundidad`

* Al final vamos a mostrar el resultado de la operacion de: `posicion * profundidad`

## Entrada de Datos

Si quieres ver la entrada de datos que me toco para practicar, se encuentra en el archivo `input_puzzle2.txt`

## Codigo

In [8]:
contenido = open('input_puzzle2.txt', 'r')
mensaje = contenido.readlines()
# comienza en posicion 0 y profundidad 0
posicion = 0
profundidad = 0
for linea in mensaje:
    comando, cantidad = linea.split(' ')[0], linea.split(' ')[1]
    if comando == 'forward':
        posicion += int(cantidad)
    if comando == 'down':
        profundidad += int(cantidad)
    if comando == 'up':
        profundidad -= int(cantidad)

total = posicion * profundidad
print(total)
    

1947824


## Resultado Parte1: 1947824

## Parte 2

--- La segunda parte ---
Según sus cálculos, el curso planeado no parece tener ningún sentido. Encuentra el manual del submarino y descubre que el proceso es en realidad un poco más complicado.

Además de la posición horizontal y la profundidad, también deberá realizar un seguimiento de un tercer valor, el objetivo, que también comienza en 0. Los comandos también significan algo completamente diferente de lo que pensó al principio:

down X aumenta tu puntería en X unidades.
up X disminuye tu puntería en X unidades.
forward X hace dos cosas:
Aumenta su posición horizontal en X unidades.
Aumenta tu profundidad por tu puntería multiplicada por X.
Una vez más, tenga en cuenta que, dado que está en un submarino, subir y bajar hacen lo contrario de lo que podría esperar: "abajo" significa apuntar en la dirección positiva.

Ahora, el ejemplo anterior hace algo diferente:

forward 5 suma 5 a su posición horizontal, un total de 5. Debido a que su objetivo es 0, su profundidad no cambia.
down 5 suma 5 a tu objetivo, lo que da como resultado un valor de 5.
forward 8 suma 8 a su posición horizontal, un total de 13. Como su objetivo es 5, su profundidad aumenta en 8 * 5 = 40.
up 3 disminuye tu puntería en 3, lo que da como resultado un valor de 2.
down 8 suma 8 a su objetivo, lo que da como resultado un valor de 10.
forward 2 suma 2 a tu posición horizontal, un total de 15. Como tu objetivo es 10, tu profundidad aumenta en 2 * 10 = 20 hasta un total de 60.
Después de seguir estas nuevas instrucciones, tendrías una posición horizontal de 15 y una profundidad de 60 (multiplicar estas produce 900).

Con esta nueva interpretación de los comandos, calcule la posición horizontal y la profundidad que tendría después de seguir el rumbo planificado. ¿Qué obtienes si multiplicas tu posición horizontal final por tu profundidad final?

## Logica de Resolucion

No hay mucha diferencia entre lo que teniamos hace un rato y ahora, solo que tenemos ahora tambien un `objetivo` y este se modificara con los comandos `up` y `down` y luego tenemosl comando `forward` que va a tener 2 acciones mencionadas en el enunciado.

## Codigo

In [10]:
contenido = open('input_puzzle2.txt', 'r')
mensaje = contenido.readlines()
# comienza en posicion 0 y profundidad 0
posicion = 0
profundidad = 0
objetivo = 0
for linea in mensaje:
    comando, cantidad = linea.split(' ')[0], linea.split(' ')[1]
    if comando == 'forward':
        posicion += int(cantidad)
        profundidad += int(cantidad) * int(objetivo)
    if comando == 'up':
        objetivo -= int(cantidad)
    if comando == 'down':
        objetivo += int(cantidad)
total = posicion * profundidad
print(total)

1813062561


## Resultado Parte2: 1813062561